src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs

```diff
@@ -52,34 +55,38 @@ public MapReduceResultsStore(ulong reduceKeyHash, MapResultsStorageType type, Tr
             }
         }
 
+        private readonly HashSet<string> _alreadyInitializedTrees = new HashSet<string>();
+
         private void InitializeTree(bool create)
         {
             var treeName = ReduceTreePrefix + _reduceKeyHash;
             var options = _tx.LowLevelTransaction.Environment.Options.RunningOn32Bits ? TreeFlags.None : TreeFlags.LeafsCompressed;
             Tree = create ? _tx.CreateTree(treeName, flags: options) : _tx.ReadTree(treeName);
 
-            ModifiedPages = new HashSet<long>();
-            FreedPages = new HashSet<long>();
+            if (!_alreadyInitializedTrees.Contains(treeName))
```

Comment:
How this can this be multiple times for the same tree instance? `InitializeTree` is supposed to be called just once

Reply:
It's used in debug endpoint only for Map-reduce visualizer purposes, so not very likely I think. Those are different instances of trees anyway so that shoudn't matter

Reply:
It gets called once per every tree. So in essence it will create a linear list of trees per batch, causing a huge slowdown and lots of wasted effort.

Reply:
Okay, I understand. You're right it might affect the performance - I haven't thought about it. But still don't understand  how `_alreadyInitializedTrees.Contains(treeName)` can return `false` as `InitializeTree ` is called once

Reply:
Edited the above comment: can return false -> can return true

Reply:
Because now there is a single HashMap per each per `MapReduceStore`. Having said that will see if that condition ever triggers, I started trying to figure out where was the multiplicative factor be generated, and could be useless.

Reply:
ok

Reply:
Is not reusing pages at all a viable strategy?